Feature: Add webpack plugin to externalize and extract script dependencies

sirreal opened the pull request:
## Description

Add a webpack plugin that externalizes WordPress script dependencies and adds a dependency manifest per entrypoint. The goal is to eliminate error-prone process of manually creating and maintaining script dependencies.

A JSON file is output alongside each entrypoint. That looks like (webpack output):

```
             ./build/annotations/index.js    9.17 KiB      39  [emitted]         annotations
                    ./build/a11y/index.js    1.78 KiB      40  [emitted]         a11y
             ./build/a11y/index.deps.json    16 bytes      40  [emitted]         a11y
      ./build/annotations/index.deps.json    56 bytes      39  [emitted]         annotations
```

More detail in the README.

~Closes~ Part of #14837 

## To do
- [x] Add tests
- [x] Update README. Document options with examples.
- [x] CHANGELOG - this package and `scripts`
- [x] Fix CI test failures
- [x] Document in handbook: [#](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#external-dependencies) [#](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/scripts.md)

## How has this been tested?

The plugin is put to use in the Gutenberg webpack config. It can be tested using `npm run build`.

<details>

<summary>Plugin output results</summary>

<!--

Produced via:

for f in $(ls build ); do echo '**'$(basename $f)'**'; echo '```js'; cat "build/$f/index.deps.json" ; echo ; echo '```'; echo; done

-->

**a11y**
```js
["wp-dom-ready"]
```

**annotations**
```js
["lodash","wp-data","wp-hooks","wp-i18n","wp-rich-text"]
```

**api-fetch**
```js
["wp-i18n","wp-url"]
```

**autop**
```js
[]
```

**blob**
```js
[]
```

**block-editor**
```js
["lodash","react","wp-a11y","wp-blob","wp-blocks","wp-components","wp-compose","wp-core-data","wp-data","wp-deprecated","wp-dom","wp-element","wp-hooks","wp-html-entities","wp-i18n","wp-is-shallow-equal","wp-keycodes","wp-rich-text","wp-token-list","wp-url","wp-viewport","wp-wordcount"]
```

**block-library**
```js
["lodash","moment","wp-a11y","wp-api-fetch","wp-autop","wp-blob","wp-block-editor","wp-blocks","wp-components","wp-compose","wp-core-data","wp-data","wp-date","wp-deprecated","wp-editor","wp-element","wp-i18n","wp-is-shallow-equal","wp-keycodes","wp-rich-text","wp-url","wp-viewport"]
```

**block-serialization-default-parser**
```js
[]
```

**block-serialization-spec-parser**
```js
[]
```

**blocks**
```js
["lodash","wp-autop","wp-blob","wp-block-serialization-default-parser","wp-compose","wp-data","wp-dom","wp-element","wp-hooks","wp-html-entities","wp-i18n","wp-is-shallow-equal","wp-shortcode"]
```

**components**
```js
["lodash","moment","react","react-dom","wp-a11y","wp-api-fetch","wp-compose","wp-dom","wp-element","wp-hooks","wp-i18n","wp-is-shallow-equal","wp-keycodes","wp-rich-text","wp-url"]
```

**compose**
```js
["lodash","wp-element","wp-is-shallow-equal"]
```

**core-data**
```js
["lodash","wp-api-fetch","wp-data","wp-deprecated","wp-url"]
```

**data**
```js
["lodash","wp-compose","wp-deprecated","wp-element","wp-is-shallow-equal","wp-priority-queue","wp-redux-routine"]
```

**date**
```js
["moment"]
```

**deprecated**
```js
["wp-hooks"]
```

**dom**
```js
["lodash"]
```

**dom-ready**
```js
[]
```

**edit-post**
```js
["lodash","wp-a11y","wp-api-fetch","wp-block-editor","wp-block-library","wp-blocks","wp-components","wp-compose","wp-core-data","wp-data","wp-editor","wp-element","wp-hooks","wp-i18n","wp-keycodes","wp-notices","wp-nux","wp-plugins","wp-url","wp-viewport"]
```

**edit-widgets**
```js
["wp-element"]
```

**editor**
```js
["lodash","react","wp-api-fetch","wp-autop","wp-blob","wp-block-editor","wp-blocks","wp-components","wp-compose","wp-core-data","wp-data","wp-date","wp-deprecated","wp-element","wp-hooks","wp-html-entities","wp-i18n","wp-keycodes","wp-notices","wp-nux","wp-rich-text","wp-url","wp-viewport","wp-wordcount"]
```

**element**
```js
["lodash","react","react-dom","wp-escape-html"]
```

**escape-html**
```js
[]
```

**format-library**
```js
["lodash","wp-block-editor","wp-components","wp-element","wp-i18n","wp-keycodes","wp-rich-text","wp-url"]
```

**hooks**
```js
[]
```

**html-entities**
```js
[]
```

**i18n**
```js
[]
```

**is-shallow-equal**
```js
[]
```

**keycodes**
```js
["lodash","wp-i18n"]
```

**list-reusable-blocks**
```js
["lodash","wp-api-fetch","wp-components","wp-compose","wp-element","wp-i18n"]
```

**notices**
```js
["lodash","wp-a11y","wp-data"]
```

**nux**
```js
["lodash","wp-components","wp-compose","wp-data","wp-element","wp-i18n"]
```

**plugins**
```js
["lodash","wp-compose","wp-element","wp-hooks"]
```

**priority-queue**
```js
[]
```

**redux-routine**
```js
["lodash"]
```

**rich-text**
```js
["lodash","wp-compose","wp-data","wp-element","wp-escape-html","wp-hooks"]
```

**shortcode**
```js
["lodash"]
```

**token-list**
```js
["lodash"]
```

**url**
```js
[]
```

**viewport**
```js
["lodash","wp-compose","wp-data"]
```

**wordcount**
```js
["lodash"]
```



</details>

## Types of changes
New feature:
Adds a new package `@wordpress/dependency-extraction-webpack-plugin`.

New feature:
Uses the plugin in `@wordpress/scripts` (will require version bump, not included initially). This can be part of another PR but is useful for testing.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

sirreal committed:
Add package

sirreal committed:
Add docs manifest

sirreal committed:
Replace hard-coded externals with webpack plugin

sirreal committed:
Update package-lock

nerrad reviewed (commented):
Nice!  

I wonder, instead of hardcoding the externals and the handle ref to use, could we make that configurable via an options array?  This would allow WordPress plugins to use this for their own externals as well.  

So the options config would be an array with two callbacks.  The first callback would be used for mapping externals, the second callback would be for mapping to the json output.  It could have the shape:

```js
const options = [
    ( request ) => { /* eternalized dependencies callback for mapping external to */ },
    ( request ) => { /* dependencies map callback */ }
];
```

This way it's somewhat extensible by WP plugins that might want to use this package for non WP externals.

This is just an idea, the actual shape of the configuration option for callbacks could likely be iterated on a bit (eg - should callbacks always be provided and the application merged?  Should the wp externals be hardcoded/permanent or can they be overridden?)

sirreal commented:
> I wonder, instead of hardcoding the externals and the handle ref to use, could we make that configurable via an options array?

Extensibility and configuration is missing at this time. I had in mind a few options that might be helpful.

Configuration for plugin extensibility allowing plugins provide script dependencies and build bundles that externalize them.

- `requestToExternal`: A function that receives module requests (`@wordpress/element`) and returns a string to externalize the module (`wp.element`):
  ```js
  const options = {
    requestToExternal: request => request === '@wordpress/element' ? 'wp.element' : undefined,
  }
  ```
- `requestToDependency`: A function that receives module requests (`@wordpress/element`) and returns the WordPress script dependency to include in the manifest list (`wp-element`):
  ```js
  const options = {
    requestToDependency: request => request === '@wordpress/element' ? 'wp-element' : undefined,
  }
  ```

I can imagine cases where options like these might be helpful as well:

- `useDefaults` - include the built-in externals like `jquery`, `lodash`, and `@wordpres/*`. This option would allow this behavior to be enabled/disabled (`true/false`).
- `injectPolyfill` - An option to include `wp-polyfill` as a dependency for all scripts.

aduth commented:
Related issue: #14837

Am I correct in thinking this should close the issue once merged?

sirreal commented:
> Related issue: #14837
> 
> Am I correct in thinking this should close the issue once merged?

Yes, thanks. I've updated the description to reflect that.

gziolo commented:
> Am I correct in thinking this should close the issue once merged?

I don't see the following file removed in this PR:
https://github.com/WordPress/gutenberg/blob/master/lib/packages-dependencies.php

If that can be easily squeezed in then sure, we can consider it as closing the issue :)

gziolo commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,119 @@
+/**
+ * WordPress dependencies
+ */
+const { camelCaseDash } = require( '@wordpress/scripts/utils' );
```
This will introduce a cyclic dependency between `@wordpess/scripts` and this package.

I think it's fine to move it to this package to resolve it.

It's used here as well:
https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L17

However, it doesn't matter where it is going to be imported from.

aduth replied:
I'm becoming quite worried about the spread of this utility, given the repeated cautions on the fact that it doesn't produce accurate results.

https://github.com/WordPress/gutenberg/pull/13814/files#r258035009

sirreal replied:
> I think it's fine to move it to this package to resolve it.

This webpack plugin would seem like a strange place to provide this util.

I _would_ like to remove the dependence on `@wordpress/scripts` from this package just to use `camelCaseDash`. It's trivial functionality, but `@wordpress/scripts` pulls in a lot of heavy packages like pupeteer.

Do folks have a clear idea of the best path forward?

gziolo replied:
Can we copy it for now? I will tackle it in the follow-up issue. I'm actually opening a new one based on feedback from @aduth where we can also discuss how to solve the issue with `escape-html` -> `escapeHTML` vs `escapeHtml` (which I personally don't quite know how to automate).

nerrad replied:
+1 to copy for now

sirreal replied:
I've copied the function to this package for now.

aduth commented:
> > Am I correct in thinking this should close the issue once merged?
> 
> I don't see the following file removed in this PR:
> https://github.com/WordPress/gutenberg/blob/master/lib/packages-dependencies.php
> 
> If that can be easily squeezed in then sure, we can consider it as closing the issue :)

I had wondered that as well, whether this would allow for this file to be removed.

gziolo commented on packages/dependency-extraction-webpack-plugin/package.json:
```diff
@@ -0,0 +1,33 @@
+	"main": "index.js",
+	"dependencies": {
+		"@wordpress/scripts": "file:../scripts",
+		"webpack": "4.8.3",
```
`crypto` is not listed as a dependency

sirreal replied:
The dependency is on Node `crypto`: https://nodejs.org/docs/latest-v10.x/api/crypto.html, not an npm package.

gziolo commented on packages/dependency-extraction-webpack-plugin/package.json:
```diff
@@ -0,0 +1,33 @@
+	"dependencies": {
+		"@wordpress/scripts": "file:../scripts",
+		"webpack": "4.8.3",
+		"webpack-sources": "1.3.0"
```
It's fine to use ranges for `webpack-sources`. I'm not quite sure why we didn't use ranges for `webpack` as well. We should fix it when we migrate to v5.0.

sirreal replied:
I've added `^` to the versions, is that sufficient or did you have something else in mind?

sirreal committed:
Extract default functions

sirreal committed:
Prepare options

sirreal committed:
Handle useDefaults for externals

sirreal committed:
Handle custom requestToExternal function

sirreal committed:
Handle custom dependency functions

sirreal committed:
Save dev dependency in README

sirreal committed:
Use prose package name as README title

sirreal committed:
Use ^ version ranges

sirreal committed:
Update return description

sirreal commented:
I've rebased, addressed some feedback, and implemented the configuration outlined in https://github.com/WordPress/gutenberg/pull/14869#issuecomment-480982372

I haven't had an opportunity to test the new functionality yet.

aduth commented:
For those following along, this pull request was discussed during today's [JavaScript Chat](https://make.wordpress.org/meetings/) ([link requires registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C5UNMSU4R/p1554818548169400

nerrad commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,171 @@
+		}
+
+		// Cascade to default if unhandled and enabled
+		if ( 'undefined' === typeof externRootRequest && this.options.useDefaults !== false ) {
```
yoda conditional, isn't this against js coding standards?  Should this be:

```suggestion
		if ( typeof externRootRequest === 'undefined' && this.options.useDefaults !== false ) {
```

sirreal replied:
Happy to make the change, I'm indifferent, but I don't see any mention of yoda conditionals in the handbook: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/

nerrad replied:
huh, I'm sure I saw a `yoda: 'off'` rule in the gb eslint rules at some point in its history but you're right, its not defined in the handbook nor is it declared (either way) in the current eslint rules.  So I guess it doesn't matter.

nerrad replied:
Although there is this section on the preferred way for type checks that applies here: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#type-checks

aduth replied:
Related: The standards revisions adopted late last year removed Yoda as being prescribed one way or another:

>Proposed Change: This guideline will be removed from the coding standards. While it is still an option for developers to use this pattern, it will not be enforced one way or the other.

https://make.wordpress.org/core/2018/12/18/proposed-revisions-to-javascript-coding-standards/

nerrad commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,171 @@
+
+	mapRequestToDependency( request ) {
+		// Handle via options.requestToDependency first
+		if ( 'function' === typeof this.options.requestToDependency ) {
```
same thing here with yoda conditional... should it be:

```suggestion
		if ( typeof this.options.requestToDependency === 'function' ) {
```

nerrad commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,171 @@
+		let externRootRequest;
+
+		// Handle via options.requestToExternal first
+		if ( 'function' === typeof this.options.requestToExternal ) {
```
yoda conditional, isn't this against js coding standards?  Should this be:

```suggestion
		if ( typeof this.options.requestToExternal === 'function' ) {
```

sirreal committed:
Use boolean value directly

sirreal committed:
Implement injectPolyfill

nerrad committed:
Invert yoda conditional

Co-Authored-By: sirreal <jon.surrell@automattic.com>

sirreal committed:
Restore regeneratorRuntime external

sirreal committed:
Extern lodash-es

sirreal committed:
Improve clarity

sirreal commented on packages/scripts/config/webpack.config.js:
```diff
@@ -81,7 +33,17 @@ const config = {
+			// externals to reduce the size of generated bundles.
+			//
+			// See: https://github.com/WordPress/gutenberg/issues/13890
+			'@babel/runtime/regenerator': 'regeneratorRuntime',
```
If I understand correctly, this should be added to resolve:

`@babel/runtime/regenerator` -> `regeneratorRuntime` (external)
and
`@babel/runtime/regenerator` -> `wp-polyfill` (script dependency)

Is that right?

https://github.com/WordPress/WordPress/blob/8ffebab3e6910123905fb238c94b7b81ef3fc1d3/wp-includes/js/dist/vendor/wp-polyfill.js#L6250

gziolo replied:
Yes, correct. We add `wp-polyfill` by default to every `wp-*` script handle but you can do it with webpack as well, good idea 😄

sirreal committed:
Handle @babel/runtime/regenerator

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -81,7 +33,6 @@ const config = {
 		filename: '[name].js',
 		path: path.resolve( process.cwd(), 'build' ),
 	},
-	externals,
```
Out of curiosity, what will happen if someone provides their own list of externals? Will it merge with what plugin provides?

sirreal replied:
This is something I want to better understand myself. I plan to add tests that will clarify the expected behavior.

I suspect that they are compatible, but they may conflict. Proper documentation —another to-do 🙂 — should indicate what to expect and how to deal with issues.

sirreal replied:
I've done my best to document this:

> **Note:** This plugin overlaps with the functionality provided by [webpack
> `externals`](https://webpack.js.org/configuration/externals). This plugin is intended to extract
> script handles from bundle compilation so that a list of script dependencies does not need to be
> manually maintained. If you don't need to extract a list of script dependencies, use the `externals`
> option directly.
> 
> This plugin is compatible with `externals`, but they may conflict. For example, adding
> `{ externals: { '@wordpress/blob': 'wp.blob' } }` to webpack configuration will effectively hide the
> `@wordpress/blob` module from the plugin and it will not be included in dependency lists.

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/README.md#webpack

I've also added some tests around this:

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/test/fixtures/with-externs/webpack.config.js#L1-L7
https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/test/fixtures/with-externs/index.js#L1-L18

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/test/__snapshots__/build.js.snap#L44-L73

gziolo commented on packages/dependency-extraction-webpack-plugin/README.md:
````diff
@@ -0,0 +1,67 @@
+Enqueue your script as usual and read the script dependencies dynamically:
+
+```php
+$script_path      = 'path/to/script.js';
````
I have a feeling that we should align spaces between variable declarations to follow strict WordPress coding styles.

gziolo commented on packages/dependency-extraction-webpack-plugin/README.md:
````diff
@@ -0,0 +1,67 @@
+}
+```
+
+Each entrypoint in the webpack bundle will include JSON file that declares the WordPress script dependencies that should be enqueued.
````
We should explain what happens to the existing `externals` when they are provided as well.

sirreal replied:
`externals` can still be used, I hope the documentation makes it clear (https://github.com/WordPress/gutenberg/pull/14869#discussion_r275461917).

gziolo reviewed (commented):
I think it looks close to being ready to merge. I added some additional comments to address before we land the very first iteration. We can make it customizable later when we get questions and requests from developers using it. It worked quite well with `wp-scripts build` so I'd recommend following the same approach.

nerrad reviewed (changes requested):
I'm pretty keen on trying out this webpack plugin in my own projects :)  However, there's still a few concerns I've noted.

- I think the documentation should be improved a bit more (as noted).
- Looks like you need `CHANGELOG.md` for all affected packages (including the new one).
- What's up with travis? It seems stuck at the moment.  I haven't seen a full green check in this pull yet so that will need resolved before merge.

sirreal committed:
Copy camelCaseDash from scripts

See https://github.com/WordPress/gutenberg/pull/14869#discussion_r274363420

sirreal commented:
I've updated the description with a to do list. As noted, the bulk of the implementation is functional and mostly complete. There's good feedback that I will address as soon as I can.

The biggest tasks are to add tests followed by updating and improving the documentation.

sirreal committed:
Add tests

sirreal committed:
Add unit tests and restructure to facilitate

Move default transform functions to a until module.
Add unit tests for transform functions.

sirreal committed:
Test external modules and dependencies file snapshot

sirreal committed:
Align adjacent `=` in README

sirreal committed:
Add options documentation to the README

sirreal committed:
Test externals conflict

sirreal committed:
Add a note about conflicts with externals

sirreal committed:
Add test case for requires

sirreal commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,193 @@
+		// Cascade to default if unhandled and enabled
+		if ( typeof externRootRequest === 'undefined' && this.options.useDefaults ) {
+			externRootRequest = defaultRequestToExternal( request );
+		}
```
I've documented this option:

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/README.md#requesttoexternal

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/README.md#requesttoexternal-and-requesttohandle

sirreal commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,193 @@
+
+		// Fall back to the request name
+		return request;
+	}
```
I've documented this option:

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/README.md#requesttohandle

https://github.com/WordPress/gutenberg/blob/1fd0b69b2cade06eb85eb3c77bc0a838d3333375/packages/dependency-extraction-webpack-plugin/README.md#requesttoexternal-and-requesttohandle

sirreal committed:
Add changelog for new package

sirreal committed:
Add @wordpress/scripts changelog and bump version

sirreal committed:
Remove describe wrapper and simplify test name

sirreal commented on packages/dependency-extraction-webpack-plugin/index.js:
```diff
@@ -0,0 +1,193 @@
+
+		case 'react-dom':
+			return 'ReactDOM';
+	}
```
I'm happy to add some documentation there. Can you point me to where you think would be appropriate?

nerrad replied:
Maybe somewhere around here: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#external-dependencies ?

simison replied:
Also https://github.com/WordPress/gutenberg/blob/master/docs/contributors/scripts.md

sirreal commented:
- I believe all feedback has been addressed
- Improved README with documentation about options and interaction with `externals`
- Added tests

I'd appreciate another round of feedback!

nerrad commented:
There still seem to be a number of fails specific to the changes in this branch before this can be reviewed?

sirreal commented:
> There still seem to be a number of automated testing fails in travis specific to the changes in this branch before this can be reviewed?

Apologies I didn't respond on this. I have looked at the e2e issues and they all appear to fail with `Error: Cannot find module '@wordpress/dependency-extraction-webpack-plugin'`. I've been unable to reproduce issues with any of the following:

- `WP_VERSION=latest DOCKER=true ./bin/run-wp-unit-tests.sh`
- `npm run test-e2e`
- `npm run test-unit-php`
- `npm run test-unit-php-multisite`

It's not clear to me why the tests wouldn't be able to find the new module. I'd appreciate if someone more familiar with the test setup could help out.

gziolo commented:
https://travis-ci.com/WordPress/gutenberg/jobs/193067071#L754

> Error: Cannot find module '@wordpress/dependency-extraction-webpack-plugin'

My bet is that you need to include `@wordpress/dependency-extraction-webpack-plugin` in the root `package.json` file to ensure it's properly linked from `node_modules` for all tools.

simison commented:
>I've been unable to reproduce issues with any of the following:

If it's not reproducible locally, I'd suspect that Travis's cache doesn't get cleaned out for changes in packages.

sirreal committed:
Try: add package to root devDeps

sirreal committed:
Update @wordpress/scripts README

gziolo commented on packages/dependency-extraction-webpack-plugin/test/fixtures/no-deps/webpack.config.js:
```diff
@@ -0,0 +1 @@
+module.exports = {};
```
To my previous point, it looks strange when there is an empty config file.

gziolo commented on packages/dependency-extraction-webpack-plugin/test/build.js:
```diff
@@ -0,0 +1,80 @@
+
+afterAll( () => rimraf( path.join( __dirname, 'build' ) ) );
+
+configFixtures.forEach( ( configCase ) => {
```
There is also `describe.each` which plays a similar role:
https://jestjs.io/docs/en/api#describeeachtable-name-fn-timeout

gziolo commented on packages/dependency-extraction-webpack-plugin/package.json:
```diff
@@ -0,0 +1,32 @@
+{
+	"name": "@wordpress/dependency-extraction-webpack-plugin",
+	"version": "1.0.0",
```
```suggestion
	"version": "1.0.0+alpha.0",
```
so we could leave room for version bump for Lerna

gziolo commented on packages/dependency-extraction-webpack-plugin/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+## 1.0.0 (tbd)
```
```suggestion
## Unreleased
```

gziolo commented on packages/scripts/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+## 3.2.0 (tbd)
```
```suggestion
## Unreleased
```

gziolo commented:
It looks like Travis is happy after you added this one line to `package.json` file. Awesome!

I left the very last list of nitpicks and I think it's ready to go. @nerrad, do you feel it's ready as well? I'm personally very excited to see how it's shaped at the moment. @sirreal did a fantastic job 🏅

sirreal committed:
Add bundling documentation to scripts

sirreal commented:
> It looks like Travis is happy after you added this one line to `package.json` file. Awesome!

I'm skeptical that this change was strictly necessary, but indeed it fixed travis so I'm happy 🙂 

I've added some documentation where it seemed most appropriate, it would be good to pay special attention to that in reviews. I'm not sure what's the most appropriate way to link between different parts of documentation, for example between package READMEs and parts of docs/.

gziolo commented:
> I'm skeptical that this change was strictly necessary, but indeed it fixed travis so I'm happy 🙂

Yup, it's probably something with how npm updates the lock file for dependencies of local packages. Every time we bump version of dependency of WordPress package in Gutenberg. We have to start the dance to enforce it gets updated in the root level of the lock file as well. My assumption is, when you put it in the root folder it gets magically fixed :)

sirreal committed:
Add rimraf before tests

sirreal commented on packages/dependency-extraction-webpack-plugin/test/build.js:
```diff
@@ -0,0 +1,80 @@
+				);
+				options.output.path = outputDirectory;
+				if ( ! options.plugins ) {
+					options.plugins = [ new DependencyExtractionWebpackPlugin() ];
```
In general it should. This saves needing to import the module and define plugins for each of the tests if we're just testing default behavior:

https://github.com/WordPress/gutenberg/blob/0d0b3bab163ee64f73461aa44a34d80652355e8c/packages/dependency-extraction-webpack-plugin/test/fixtures/wordpress-require/webpack.config.js#L1

Happy to remove this and add the plugin explicitly if you prefer.

sirreal replied:
I see https://github.com/WordPress/gutenberg/pull/14869#discussion_r275706522, I'll make them explicit 🙂

sirreal committed:
Revert @wordpress/scripts version bump

sirreal committed:
Use alpha version suffix

gziolo committed:
Use "unreleased" versions in changelogs

Co-Authored-By: sirreal <jon.surrell@automattic.com>

sirreal committed:
Use the plugin explicitly

sirreal committed:
Use describe.each over explicit loop

sirreal committed:
Use before/afterEach, add comment about cleanup

sirreal commented:
I believe all feedback from https://github.com/WordPress/gutenberg/pull/14869#pullrequestreview-227030584 has been addressed.

gziolo commented on packages/scripts/README.md:
```diff
@@ -271,17 +271,9 @@ The `build` and `start` commands use [webpack](https://webpack.js.org/) behind t
 * [Output](https://webpack.js.org/configuration/output): `build/index.js`
-* [Externals](https://webpack.js.org/configuration/externals). These are libraries that are to be found in the global scope:
-
-Package | Input syntax | Output
```
Can we keep this list of externals but move it to the webpack plugin docs?

sirreal replied:
I added the following:

By default, the following module requests are handled:

> | Request | Global | Script handle |
> | --- | --- | --- |
> | `@babel/runtime/regenerator` | `regeneratorRuntime` | `wp-polyfill` |
> | `@wordpress/*` | `wp.*` | `wp-*` |
> | `jquery` | `jQuery` | `jquery` |
> | `jquery` | `jQuery` | `jquery` |
> | `lodash-es` | `lodash` | `lodash` |
> | `lodash` | `lodash` | `lodash` |
> | `moment` | `moment` | `moment` |
> | `react-dom` | `ReactDOM` | `react-dom` |
> | `react` | `React` | `react` |

gziolo commented on packages/scripts/README.md:
```diff
@@ -271,17 +271,9 @@ The `build` and `start` commands use [webpack](https://webpack.js.org/) behind t
-lodash-es | `import x from 'lodash-es';` | `var x = window.lodash.x;`
-WordPress packages | `import x from '@wordpress/package-name'` | `var x = window.wp.packageName.x`
+* [Plugins](https://webpack.js.org/configuration/plugins): The webpack plugin provided by
+[`@wordpress/dependency-extraction-webpack-plugin`](../dependency-extraction-webpack-plugin) is used
```
```suggestion
[`@wordpress/dependency-extraction-webpack-plugin`](/packages/dependency-extraction-webpack-plugin/README.md) is used
```
to ensure it links properly on npm and in Gutenberg handbook

gziolo commented on packages/scripts/README.md:
```diff
@@ -271,17 +271,9 @@ The `build` and `start` commands use [webpack](https://webpack.js.org/) behind t
-WordPress packages | `import x from '@wordpress/package-name'` | `var x = window.wp.packageName.x`
+* [Plugins](https://webpack.js.org/configuration/plugins): The webpack plugin provided by
+[`@wordpress/dependency-extraction-webpack-plugin`](../dependency-extraction-webpack-plugin) is used
+with the default configuration.
```
```suggestion
with the default configuration. It ensures that existing WordPress scripts aren't included in the build file generated.
```

We could also add some short clarification about what it does.

nerrad reviewed (approved):
Let's 🚢 it!  Awesome work!

gziolo reviewed (approved):
This PR tests well, I didn't notice more significant changes before the old and new output of production build files. There is one subtle difference in the build file but it seems to be legit:

_old_

```js
194:function(e,t){!function(){e.exports=this.wp.domReady}()},
```
_new_

```js
194:function(e,t){e.exports=wp.domReady},
```

the same applies for a development build:

_old_

```js
/***/ "@wordpress/dom-ready":
/*!*******************************************!*\
  !*** external {"this":["wp","domReady"]} ***!
  \*******************************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["domReady"]; }());

/***/ })
```

_new_

```js
/***/ "@wordpress/dom-ready":
/*!******************************!*\
  !*** external "wp.domReady" ***!
  \******************************/
/*! no static exports found */
/***/ (function(module, exports) {

module.exports = wp.domReady;

/***/ })
```

gziolo committed:
Update README href

Co-Authored-By: sirreal <jon.surrell@automattic.com>

sirreal committed:
Add externals info to scripts README

sirreal committed:
Add default module request handling table to README

sirreal committed:
Use this externalType and array object path

sirreal committed:
Update tests to use array paths

sirreal committed:
Add dynamic-import test

sirreal commented:
Rebased.

> This PR tests well, I didn't notice more significant changes before the old and new output of production build files. There is one subtle difference in the build file but it seems to be legit:

This review prompted me to reconsider the current approach and take some time to [review](https://github.com/webpack/webpack/blob/b72b96a2bc5d2c7a308d0943567e0db191d3c358/lib/ExternalsPlugin.js) [webpack](https://github.com/webpack/webpack/blob/b72b96a2bc5d2c7a308d0943567e0db191d3c358/lib/ExternalModuleFactoryPlugin.js) [internals](https://github.com/webpack/webpack/blob/b72b96a2bc5d2c7a308d0943567e0db191d3c358/lib/ExternalModule.js).

I've pushed a small change to use `this`-based externals. Here are a few examples of current output:

```
/***/ "@wordpress/dom-ready":
/*!*******************************************!*\
  !*** external {"this":["wp","domReady"]} ***!
  \*******************************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["domReady"]; }());

/***/ })

/***/ "lodash":
/*!**********************************!*\
  !*** external {"this":"lodash"} ***!
  \**********************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["lodash"]; }());

/***/ })
```

Notably, object paths may be provided as arrays of strings, for example `@wordpress/i18n` is now `[ 'wp', 'i18n' ]`. which results in more robust output `this["wp"]["i18n"]` compared to `wp.domReady`. Optimization may still result in the latter, but producing `wp.${ packageName }` directly could result in invalid output.

With these finishing touches, I'm satisfied with this initial implementation 🙂

sirreal commented:
There is a failing e2e test:

>  FAIL  specs/undo.test.js (8.347s)
>  ● undo › Should undo to expected level intervals

The same test appears to be failing on most recent branches so I suspect the failure is unrelated.

gziolo commented:
> The same test appears to be [failing](https://travis-ci.com/WordPress/gutenberg/jobs/193451382) on [most](https://travis-ci.com/WordPress/gutenberg/jobs/193437574) [recent](https://travis-ci.com/WordPress/gutenberg/jobs/193434167) [branches](https://travis-ci.com/WordPress/gutenberg/jobs/193398214) so I suspect the failure is unrelated.

Yes, nothing to worry about. I restarted the job which had the failure. There are a few tests which fail from time to time and we couldn't discover the reason for months now 🤷‍♂️

gziolo commented:
This is how it looks after changes applied:

_development_

```js
/***/ "@wordpress/dom-ready":
/*!*******************************************!*\
  !*** external {"this":["wp","domReady"]} ***!
  \*******************************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["domReady"]; }());

/***/ })
```

_production_
```js
194:function(e,t){!function(){e.exports=this.wp.domReady}()},
```

It looks like it's identical to what we have in `master`. Awesome. Let's get this in! 💃

gziolo commented:
We made it happen, awesome work @sirreal. It will open new ways for optimizing how JS dependencies are managed for plugin authors and WordPress core as well.

🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

nerrad commented:
@gziolo any idea on when this will be published?

gziolo commented:
> @gziolo any idea on when this will be published?

Whenever WordPress trunk is open for patches after WordPress 5.2 release process is close to ready? We might wait until plugin release. I hope this will happen at the end of the month the latest.

youknowriad commented:
Great work @sirreal  👍 

@nerrad I expect this to be published once Core `trunk` branch is open for business. At that point, we could create a `wp/5.2` branch here and backport the last Gutenberg release into `wp/trunk`. I think by the time the next Gutenberg version is released (in two weeks), `trunk` will probably be open for features...
